proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1097 @@
+
+#. Introduce ``-XScopedForAlls``. With ``-XScopedForAlls``, any type variables
+   mentioned in an explicit ``forall`` scopes over an expression. This applies
+   to the following constructs:
+
+   * Function bindings
+   * Pattern synonym bindings (including in any ``where`` clause)
+   * Expression type signatures
+
+   (Alternative name: ``-XExtendedForAllScope``. See `Vlad's comment <https://github.com/ghc-proposals/ghc-proposals/pull/448#discussion_r738276607>`_.).
+
+   Separating out ``-XScopedForAlls`` gets us closer to the CSP_.
+
+#. The extension ``-XScopedTypeVariables`` would imply all of the above
+   extensions; this way, ``-XScopedTypeVariables`` does not change from its
+   current meaning.
+
+#. Introduce ``-XImplicitForAll``, on by default. With ``-XImplicitForAll``,
+   an out-of-scope type variable mentioned in various constructs (listed below)
+   is implicitly brought into scope over the construct. With ``-XNoImplicitForAll``,
+   this implicit scoping does not happen, and the use of the variable is an error.
+
+   Constructs affected:
+
+   1. Type signatures for variable declarations, methods, and foreign imports & exports.
+      Example: ``let f :: a -> a; f = ... in ...`` becomes
+      ``let f :: forall a. a -> a; f = ... in ...``.
+
+   #. Kind signatures. Example: ``type T :: k -> Type`` becomes ``type T :: forall k. k -> Type``.
+
+   #. GADT constructor declarations. Example: ``MkG :: a -> Maybe b -> G (Either Int b)``
+      becomes ``MkG :: forall a b. a -> Maybe b -> G (Either Int b)``.
+
+   #. Pattern synonym signatures. Example: ``pattern P :: a -> Maybe a`` becomes
+      ``pattern P :: forall a. a -> Maybe a``. Implicit quantification in pattern synonyms
+      always produces *universal* variables, never existential ones.
+
+   #. Type annotations in expressions and ``SPECIALISE`` pragmas. Example:
+      ``Right True :: Either a Bool`` becomes ``Right True :: forall a. Either a Bool``.
+
+   #. Types in a ``deriving`` clause. Example: ``data T deriving (C a)`` becomes
+      ``data T deriving (forall a. C a)``.
+
+   #. Instance heads, including standalone-deriving instances.
+      Example: ``instance Show a => Show (Maybe a)`` becomes
+      ``instance forall a. Show a => Show (Maybe a)``.
+
+   #. Type and data family instances, as well as closed type family equations.
+      Example: ``type instance F (Maybe a) = Int``
+      becomes ``type instance forall a. F (Maybe a) = Int``.
+
+   #. ``RULES`` pragmas.
+      Example: ``{-# RULES "name" forall (x :: Maybe a). foo x = 5 #-}``
+      becomes ``{-# RULES "name" forall a. forall (x :: Maybe a). foo x = 5 #-}``.
+      (The double-\ ``forall`` syntax separates type variables like ``a`` from
+      term variables like ``x``.)
+
+   This extension is part of accepted, unimplemented proposal `#285`_; the only change is including ``RULES`` pragmas, which @Ericson2314 simplify forgot to include in `#285`_ (his own admission).
```

Comment:
This extension seems quite orthogonal to the rest of the proposal. If the goal is to make a trivial amendment to an existing proposal, better make a PR, and it'll be expedited quickly.

Reply:
The subtle interaction is that we can turn off `ScopedTypeVariables` without making special cases for RULES or making certain things impossible to express.

Now it may not seem like `ScopedTypeVariables` has much to do with this, but the docs do mention this: https://downloads.haskell.org/ghc/latest/docs/html/users_guide/exts/rewrite_rules.html?highlight=ScopedTypeVariables . the problem is the current docs conceive of `ScopedTypeVariables` as making type variables "have scopes", but the more modern framing is that type variables always have scopes, and it is just a question of *what scope* they are bound in.

------

All that terribly confusing stuff aside, I support making a separate PR for this. I also support just doing #285 too as a prereq for the same reason -- it is also "bug fix" in my mind. If people are willing to treat the little bug fix in isolation, that's great.

Then, this proposal can still say "oh actually, that little bug fix in fact fits into this much larger context ....", somewhat belying the fact that it was a little bug fix, but at least this proposal just needs to discuss the little bugfix as part of the motivation, and not as part of the proposed change section.

Reply:
I opened https://github.com/ghc-proposals/ghc-proposals/pull/480 doing this.

Reply:
I was asked to include #285 in this mega-proposal. Though this was not clear at the time, I think it absolutely makes sense here (even though this part of this proposal recapitulates an existing proposal), precisely because it's (part of) what delivers the Explicit Binding Principle.